lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```diff
@@ -0,0 +1,172 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Assists a class in presenting a synchronous interface that is implemented
+ * via asynchronous interfaces and events. When a method call is received
+ * by the interface, parameter values are captured and the initiation of
+ * asynchronous processing is encapsulated in a callable object. When
+ * {@code block()} is called with the internal interface identifier, the
+ * lock is taken, the asynchronous processing is initiated, and the caller
+ * is put to sleep. After all of the asynchronous processing is complete the
+ * caller is released with a call to {@code release()}.
+ */
+public final class MultiAsyncToSync {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSync.class.getName());
+
+  private final ConcurrentLinkedQueue<ComplexCondition> freeQueue = new ConcurrentLinkedQueue<>();
+  private final ConcurrentHashMap<Long, ComplexCondition> sleeperMap = new ConcurrentHashMap<>();
+  private final long timeoutPeriod;
+  private final TimeUnit timeoutUnits;
+
+  /**
+   * Initialize a multiple asynchronous to synchronous object with a specified timeout value.
+   * @param timeoutPeriod The length of time in units given by the the timeoutUnits
+   *                      parameter before the condition automatically times out.
+   * @param timeoutUnits The unit of time for the timeoutPeriod parameter.
+   */
+  public MultiAsyncToSync(final long timeoutPeriod, final TimeUnit timeoutUnits) {
+    this.timeoutPeriod = timeoutPeriod;
+    this.timeoutUnits = timeoutUnits;
+  }
+
+  /**
+   * Put the caller to sleep on a specific release identifier.
+   * @param identifier The identifier required to awake the caller via the {@code release()} method.
+   * @param asyncProcessor A {@code FutureTask} object which returns {@code TAsync} that initiates the asynchronous
+   *                       processing associated with the call. This will occur inside the condition lock
+   *                       to prevent the processing from generating the signal before the calling thread blocks.
+   *                       Error conditions should be handled by throwing an exception which the caller
+   *                       will catch.
+   * @param <TAsync> The return type of the {@code asyncProcessor};
+   * @return A boolean value that indicates whether or not a timeout or error occurred.
+   * @throws InterruptedException The thread was interrupted while waiting on a condition.
+   * @throws InvalidIdentifierException The identifier parameter is invalid.
+   * @throws Exception The future task object referenced by the {@code asyncProcessor}
+   *                   parameter threw an exception.
+   */
+  public <TAsync> boolean block(final long identifier, final FutureTask<TAsync> asyncProcessor) throws Exception {
```

Comment:
Be more specific with the exception classes in the `throws` clause. (i.e. use `InterruptedException` etc.)

Reply:
Done. Intellij greys out the specific exceptions are in a list with Exception.
     `throws InterruptedException, InvalidIdentifierException, Exception {`
However the code quality checks still pass.